File: src/main/java/me/despical/kotl/handler/ChatManager.java
/*
 * KOTL - Don't let others to climb top of the ladders!
 * Copyright (C) 2021 Despical and contributors
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

package me.despical.kotl.handler;

import me.despical.commons.configuration.ConfigUtils;
import me.despical.commons.util.Strings;
import org.apache.commons.lang.StringUtils;
import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.entity.Player;

import me.clip.placeholderapi.PlaceholderAPI;
import me.despical.kotl.Main;
import me.despical.kotl.arena.Arena;

import java.util.List;

/**
 * @author Despical
 * <p>
 * Created at 22.06.2020
 */
public class ChatManager {

	private String prefix;
	
	private final Main plugin;
	private FileConfiguration config;
	
	public ChatManager(Main plugin) {
		this.plugin = plugin;
		this.config = ConfigUtils.getConfig(plugin, "messages");
		this.prefix = message("In-Game.Plugin-Prefix");
	}
	
	public String getPrefix() {
		return prefix;
	}

	public String coloredRawMessage(String message) {
		return Strings.format(message);
	}

	public String prefixedRawMessage(String message) {
		return prefix + coloredRawMessage(message);
	}

	public String message(String path) {
		return coloredRawMessage(config.getString(path));
	}

	public String prefixedMessage(String path) {
		return prefix + message(path);
	}
	
	public String message(String path, Player player) {
		String returnString = config.getString(path);
		returnString = StringUtils.replace(returnString, "%player%", player.getName());

		if (plugin.getServer().getPluginManager().isPluginEnabled("PlaceholderAPI")) {
			returnString = PlaceholderAPI.setPlaceholders(player, returnString);
		}

		return coloredRawMessage(returnString);
	}

	private String formatMessage(Arena arena, String message, Player player) {
		String returnString = message;

		returnString = StringUtils.replace(returnString, "%player%", player.getName());
		returnString = formatPlaceholders(returnString, arena);

		if (plugin.getServer().getPluginManager().isPluginEnabled("PlaceholderAPI")) {
			returnString = PlaceholderAPI.setPlaceholders(player, returnString);
		}

		return coloredRawMessage(returnString);
	}

	private String formatPlaceholders(String message, Arena arena) {
		String returnString = message;

		returnString = StringUtils.replace(returnString, "%arena%", arena.getId());
		returnString = StringUtils.replace(returnString, "%players%", Integer.toString(arena.getPlayers().size()));
		returnString = StringUtils.replace(returnString, "%king%", arena.getKingName());
		return returnString;
	}

	public List<String> getStringList(String path) {
		return config.getStringList(path);
	}
	
	public void broadcastAction(Arena arena, Player player, ActionType action) {
		String path;

		switch (action) {
			case JOIN:
				path = "In-Game.Join";
				break;
			case LEAVE:
				path = "In-Game.Leave";
				break;
			case NEW_KING:
				path = "In-Game.New-King";

				arena.getHologram().appendLine(formatMessage(arena, message("In-Game.Last-King-Hologram"), player));
				break;
			default:
				return;
		}

		arena.broadcastMessage(prefix + formatMessage(arena, message(path), player));
	}
	
	public void reloadConfig() {
		config = ConfigUtils.getConfig(plugin, "messages");
		prefix = message("In-Game.Plugin-Prefix");
	}

	public enum ActionType {
		JOIN, LEAVE, NEW_KING
	}
}

File: src/main/java/me/despical/kotl/handler/ChatManager.kt
package me.despical.kotl.handler

import me.clip.placeholderapi.PlaceholderAPI
import me.despical.commonsbox.configuration.ConfigUtils
import me.despical.kotl.Main
import me.despical.kotl.arena.Arena
import org.apache.commons.lang.StringUtils
import org.bukkit.ChatColor
import org.bukkit.configuration.file.FileConfiguration
import org.bukkit.entity.Player

/**
 * @author Despical
 * <p>
 * Created at 31.07.2020
 */
class ChatManager constructor(val plugin: Main) {

	private var config: FileConfiguration = ConfigUtils.getConfig(plugin, "messages")
	private val prefix: String = colorRawMessage(config.getString("In-Game.Plugin-Prefix"))

	fun getPrefix(): String = prefix

	fun colorRawMessage(message: String): String = ChatColor.translateAlternateColorCodes('&', message)

	fun colorMessage(message: String): String = ChatColor.translateAlternateColorCodes('&', config.getString(message))

	fun colorMessage(message: String, player: Player): String {
		var returnString: String = config.getString(message)
		if (plugin.server.pluginManager.isPluginEnabled("PlaceholderAPI")) {
			returnString = PlaceholderAPI.setPlaceholders(player, returnString)
		}
		return ChatColor.translateAlternateColorCodes('&', returnString)
	}

	private fun formatMessage(arena: Arena, message: String, player: Player): String {
		var returnString: String = message
		returnString = StringUtils.replace(returnString, "%player%", player.name)
		returnString = colorRawMessage(formatPlaceholders(returnString, arena))
		if (plugin.server.pluginManager.isPluginEnabled("PlaceholderAPI")) {
			returnString = PlaceholderAPI.setPlaceholders(player, returnString)
		}
		return returnString
	}

	private fun formatPlaceholders(message: String, arena: Arena): String {
		var returnString: String = message
		returnString = StringUtils.replace(returnString, "%arena%", arena.id)
		returnString = StringUtils.replace(returnString, "%players%", arena.players.size.toString())
		returnString = StringUtils.replace(returnString, "%king%", if (arena.king == null) "Nobody" else arena.king!!.name)
		return returnString
	}

	fun broadcastMessage(a: Arena, msg: String) {
		for (p in a.players) {
			p.sendMessage(prefix + msg)
		}
	}

	fun broadcastAction(a: Arena, p: Player, action: ActionType) {
		val message: String
		when (action) {
			ActionType.NEW_KING -> {
				message = formatMessage(a, colorMessage("In-Game.New-King"), p)
				a.hologram.getLine(0).removeLine()
				a.hologram.insertTextLine(0, formatMessage(a, colorMessage("In-Game.Last-King-Hologram"), p))
			}
			ActionType.JOIN -> message = formatMessage(a, colorMessage("In-Game.Join"), p)
			ActionType.LEAVE -> message = formatMessage(a, colorMessage("In-Game.Leave"), p)
		}
		for (player in a.players) {
			player.sendMessage(prefix + message)
		}
	}

	fun reloadConfig() {
		config = ConfigUtils.getConfig(plugin, "messages")
	}

	enum class ActionType {
		JOIN, LEAVE, NEW_KING
	}
}